File: research/experiments/sweep/runs/mathexpr-full-kotlin-sonnet45/src/main/kotlin/mathexpr/Parser.kt
package mathexpr

class Parser(private val tokens: List<Token>) {
    private var pos = 0

    private fun peek(): Token? {
        return if (pos < tokens.size) tokens[pos] else null
    }

    private fun consume(): Token {
        if (pos >= tokens.size) {
            throw IllegalArgumentException("Unexpected end of input")
        }
        return tokens[pos++]
    }

    private fun expect(kind: TokenKind): Token {
        if (pos >= tokens.size) {
            throw IllegalArgumentException("Expected ${kind.name.lowercase()}")
        }
        val token = tokens[pos++]
        if (token.kind != kind) {
            throw IllegalArgumentException("Expected ${kind.name.lowercase()} but got ${token.kind.name.lowercase()}")
        }
        return token
    }

    fun parseExpression(): AstNode {
        val result = parseAddSub()
        if (pos < tokens.size) {
            throw IllegalArgumentException("Unexpected token after expression")
        }
        return result
    }

    private fun parseAddSub(): AstNode {
        var left = parseMulDiv()
        while (true) {
            val token = peek()
            if (token == null || (token.kind != TokenKind.PLUS && token.kind != TokenKind.MINUS)) {
                break
            }
            consume()
            val op = if (token.kind == TokenKind.PLUS) "+" else "-"
            val right = parseMulDiv()
            left = binaryExpr(op, left, right)
        }
        return left
    }

    private fun parseMulDiv(): AstNode {
        var left = parsePower()
        while (true) {
            val token = peek()
            if (token == null || (token.kind != TokenKind.STAR && token.kind != TokenKind.SLASH && token.kind != TokenKind.PERCENT)) {
                break
            }
            consume()
            val op = when (token.kind) {
                TokenKind.STAR -> "*"
                TokenKind.SLASH -> "/"
                TokenKind.PERCENT -> "%"
                else -> throw IllegalStateException("Unexpected token kind")
            }
            val right = parsePower()
            left = binaryExpr(op, left, right)
        }
        return left
    }

    private fun parsePower(): AstNode {
        val left = parseUnary()
        val token = peek()
        if (token != null && token.kind == TokenKind.POWER) {
            consume()
            val right = parsePower() // Right-associative: recurse into same level
            return binaryExpr("**", left, right)
        }
        return left
    }

    private fun parseUnary(): AstNode {
        val token = peek()
        if (token != null && token.kind == TokenKind.MINUS) {
            consume()
            val operand = parseUnary() // Allow chaining: --5
            return unaryExpr("-", operand)
        }
        return parseAtom()
    }

    private fun parseAtom(): AstNode {
        val token = consume()
        when (token.kind) {
            TokenKind.NUMBER -> {
                val value = token.value.toDouble()
                return numberLiteral(value)
            }
            TokenKind.LPAREN -> {
                val expr = parseAddSub()
                expect(TokenKind.RPAREN)
                return expr
            }
            else -> throw IllegalArgumentException("Unexpected token: ${token.kind.name.lowercase()} '${token.value}'")
        }
    }
}

fun parse(tokens: List<Token>): AstNode {
    if (tokens.isEmpty()) {
        throw IllegalArgumentException("Unexpected end of input")
    }
    val parser = Parser(tokens)
    return parser.parseExpression()
}

File: research/experiments/sweep/runs/mathexpr-full-cpp-sonnet45/src/parser.cpp
#include "parser.h"
#include <stdexcept>

class Parser {
private:
    const std::vector<Token>& tokens;
    size_t pos;

    const Token* peek() const {
        if (pos >= tokens.size()) {
            return nullptr;
        }
        return &tokens[pos];
    }

    void consume() {
        if (pos < tokens.size()) {
            pos++;
        }
    }

    std::shared_ptr<AstNode> parseAtom() {
        const Token* token = peek();
        if (!token) {
            throw std::runtime_error("Unexpected end of input");
        }

        if (token->kind == TokenKind::Number) {
            double value = std::stod(token->value);
            consume();
            return numberLiteral(value);
        }

        if (token->kind == TokenKind::LParen) {
            consume();
            auto expr = parseAddSub();
            const Token* rparen = peek();
            if (!rparen || rparen->kind != TokenKind::RParen) {
                throw std::runtime_error("Expected rparen");
            }
            consume();
            return expr;
        }

        throw std::runtime_error("Unexpected token");
    }

    std::shared_ptr<AstNode> parseUnary() {
        const Token* token = peek();
        if (token && token->kind == TokenKind::Minus) {
            consume();
            auto operand = parseUnary();
            return unaryExpr("-", operand);
        }
        return parseAtom();
    }

    std::shared_ptr<AstNode> parsePower() {
        auto left = parseUnary();
        const Token* token = peek();
        if (token && token->kind == TokenKind::Power) {
            consume();
            auto right = parsePower();  // Right-associative: recurse same level
            return binaryExpr("**", left, right);
        }
        return left;
    }

    std::shared_ptr<AstNode> parseMulDiv() {
        auto left = parsePower();
        while (true) {
            const Token* token = peek();
            if (!token) break;

            std::string op;
            if (token->kind == TokenKind::Star) {
                op = "*";
            } else if (token->kind == TokenKind::Slash) {
                op = "/";
            } else if (token->kind == TokenKind::Percent) {
                op = "%";
            } else {
                break;
            }

            consume();
            auto right = parsePower();
            left = binaryExpr(op, left, right);
        }
        return left;
    }

    std::shared_ptr<AstNode> parseAddSub() {
        auto left = parseMulDiv();
        while (true) {
            const Token* token = peek();
            if (!token) break;

            std::string op;
            if (token->kind == TokenKind::Plus) {
                op = "+";
            } else if (token->kind == TokenKind::Minus) {
                op = "-";
            } else {
                break;
            }

            consume();
            auto right = parseMulDiv();
            left = binaryExpr(op, left, right);
        }
        return left;
    }

public:
    Parser(const std::vector<Token>& tokens) : tokens(tokens), pos(0) {}

    std::shared_ptr<AstNode> parse() {
        auto result = parseAddSub();
        if (pos < tokens.size()) {
            throw std::runtime_error("Unexpected token after expression");
        }
        return result;
    }
};

std::shared_ptr<AstNode> parse(const std::vector<Token>& tokens) {
    Parser parser(tokens);
    return parser.parse();
}
